File: services/backend/src/agent_cloud_backend/models/scheduled_task.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Text, func, text
from sqlalchemy.orm import Mapped, mapped_column

from agent_cloud_backend.models.base import Base, TimestampMixin, uuid_pk


class ScheduledTask(Base, TimestampMixin):
    __tablename__ = "scheduled_tasks"

    id: Mapped[uuid.UUID] = uuid_pk()
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    agent_config_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("agent_configs.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(Text, nullable=False)
    prompt: Mapped[str] = mapped_column(Text, nullable=False)
    schedule_kind: Mapped[str] = mapped_column(nullable=False)  # once | interval | cron
    schedule_expr: Mapped[str] = mapped_column(nullable=False)
    schedule_tz: Mapped[str] = mapped_column(default="Asia/Shanghai", nullable=False)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    next_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    running_since: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_status: Mapped[str | None] = mapped_column(nullable=True)  # ok | error | skipped
    last_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    last_delivery_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    last_run_session_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("sessions.id", ondelete="SET NULL"), nullable=True
    )
    origin_session_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("sessions.id", ondelete="SET NULL"), nullable=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    __table_args__ = (
        Index(
            "ix_scheduled_tasks_due",
            "enabled",
            "next_run_at",
            postgresql_where=text("next_run_at IS NOT NULL"),
        ),
    )

File: apps/mobile/lib/models/scheduled_task.dart
/// 值守任务(定时任务):backend scheduled_tasks 的 app 侧模型。
class ScheduledTask {
  final String id;
  final String agentConfigId;
  final String name;
  final String prompt;
  final String scheduleKind; // once | interval | cron
  final String scheduleExpr;
  final bool enabled;
  final DateTime? nextRunAt;
  final DateTime? lastRunAt;
  final String? lastStatus; // ok | error | skipped
  final String? lastError;

  const ScheduledTask({
    required this.id,
    required this.agentConfigId,
    required this.name,
    required this.prompt,
    required this.scheduleKind,
    required this.scheduleExpr,
    required this.enabled,
    this.nextRunAt,
    this.lastRunAt,
    this.lastStatus,
    this.lastError,
  });

  factory ScheduledTask.fromJson(Map<String, dynamic> j) => ScheduledTask(
        id: j['id'] as String,
        agentConfigId: j['agent_config_id'] as String,
        name: j['name'] as String,
        prompt: j['prompt'] as String,
        scheduleKind: j['schedule_kind'] as String,
        scheduleExpr: j['schedule_expr'] as String,
        enabled: j['enabled'] as bool? ?? true,
        nextRunAt: _ts(j['next_run_at']),
        lastRunAt: _ts(j['last_run_at']),
        lastStatus: j['last_status'] as String?,
        lastError: j['last_error'] as String?,
      );

  static DateTime? _ts(dynamic v) =>
      v is String ? DateTime.tryParse(v)?.toLocal() : null;

  /// 周期的人话描述(列表副行)。
  String get scheduleLabel => switch (scheduleKind) {
        'cron' => 'Cron:$scheduleExpr',
        'interval' => '每 ${_intervalHuman(scheduleExpr)}',
        'once' => '一次性',
        _ => scheduleExpr,
      };

  /// 后端把 interval 归一化成纯秒("30m"→"1800"),展示时转回人话。
  static String _intervalHuman(String expr) {
    final secs = int.tryParse(expr);
    if (secs == null || secs <= 0) return expr; // "30m" 等带单位形式原样
    if (secs % 86400 == 0) return '${secs ~/ 86400} 天';
    if (secs % 3600 == 0) return '${secs ~/ 3600} 小时';
    if (secs % 60 == 0) return '${secs ~/ 60} 分钟';
    return '$secs 秒';
  }
}
